subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java

```diff
@@ -72,18 +140,32 @@ private boolean isAnnotationType(String[] interfaces) {
 
     @Override
     public FieldVisitor visitField(int access, String name, String desc, String signature, Object value) {
-        if (isConstant(access) && !isPrivate(access) && value!=null && constants != null) {
+        maybeAddDependentType(descTypeOf(desc));
+        if (isConstant(access) && !isPrivate(access) && value != null && constants != null) {
```

Comment:
maybe extract this into an `isAccessibleConstant`  method and remove the comment below